File: backend/app/services/downloader.py
import os
import yt_dlp
from typing import Optional

class DownloaderService:
    @staticmethod
    def get_youtube_audio(url: str, output_path: str) -> Optional[str]:
        """
        Downloads the best audio from a YouTube URL and converts it to mp3.
        Returns the path to the downloaded file.
        """
        ydl_opts = {
            'format': 'bestaudio/best',
            'postprocessors': [{
                'key': 'FFmpegExtractAudio',
                'preferredcodec': 'mp3',
                'preferredquality': '192',
            }],
            'outtmpl': f'{output_path}/%(id)s.%(ext)s',
            'quiet': True,
        }
        
        try:
            with yt_dlp.YoutubeDL(ydl_opts) as ydl:
                info = ydl.extract_info(url, download=True)
                return f"{output_path}/{info['id']}.mp3"
        except Exception as e:
            print(f"Error downloading YouTube audio: {e}")
            return None

downloader_service = DownloaderService()

File: backend/services/downloader.js
const youtubedl = require('youtube-dl-exec');
const path = require('path');
const fs = require('fs');

class DownloaderService {
  async getYoutubeAudio(url, outputDir) {
    if (!fs.existsSync(outputDir)) {
      fs.mkdirSync(outputDir, { recursive: true });
    }

    try {
      // Get video info to get ID
      const info = await youtubedl(url, {
        dumpSingleJson: true,
        noCheckCertificates: true,
        noWarnings: true,
        preferFreeFormats: true,
        addHeader: ['referer:youtube.com', 'user-agent:googlebot']
      });

      const videoId = info.id;
      const outputPath = path.join(outputDir, `${videoId}.mp3`);

      // Download audio
      await youtubedl(url, {
        extractAudio: true,
        audioFormat: 'mp3',
        audioQuality: 0,
        output: path.join(outputDir, `${videoId}.%(ext)s`),
      });

      return outputPath;
    } catch (error) {
      console.error("Error in YouTube download:", error);
      return null;
    }
  }
}

module.exports = new DownloaderService();
